[WIP] AWS compatibility for federation cluster and e2e

colhom opened the pull request:
I've been testing this and have reached a point where the e2e tests run, and some test failures are popping up which are not overtly related to AWS specific things.

``` sh
SSSSSSSSSSSSSSSS

Summarizing 5 Failures:

[Fail] [k8s.io] [Feature:Federation] Federated Services DNS [BeforeEach] should be able to discover a federated service 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:233

[Fail] [k8s.io] [Feature:Federation] Federated Services Service creation [It] should create matching services in underlying clusters 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:233

[Fail] [k8s.io] Federated ingresses [Feature:Federation] Federated Ingresses [It] should create and update matching ingresses in underlying clusters 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federated-ingress.go:289

[Fail] [k8s.io] [Feature:Federation] Federated Services DNS [BeforeEach] non-local federated service [Slow] missing local service should never find DNS entries for a missing local service 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:233

[Fail] [k8s.io] [Feature:Federation] Federated Services DNS [BeforeEach] non-local federated service should be able to discover a non-local federated service 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:233

Ran 16 of 383 Specs in 519.872 seconds
FAIL! -- 11 Passed | 5 Failed | 1 Pending | 366 Skipped --- FAIL: TestE2E (519.89s)
```

\cc @quinton-hoole @madhusudancs for advice. Should I investigate further?

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31908)

<!-- Reviewable:end -->

Commit:
federation: aws compatibility for cluster and e2e frameworks

madhusudancs commented:
> \cc @quinton-hoole @madhusudancs for advice. Should I investigate further?

I can help you investigate this if you give me access to the logs or email them to me.

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[build/release.sh, line 34 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQbjPYTGCBxRnRkWsHU:-KQbjPYTGCBxRnRkWsHV:bp1kzn1) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/build/release.sh#L34)):*
> ```Shell
> kube::build::verify_prereqs
> 
> if [[ $KUBE_RELEASE_DO_COMPILE =~ ^[yY]$ ]]; then
> ```

I am not entirely sure about this change, particularly about putting a large portion of this script behind an environment variable. What stops us from having a separate script that just has

```
kube::release::package_tarballs
kube::release::package_hyperkube
```
?

@david-mcmahon @thockin for their opinions.

---

*[cluster/kube-util.sh, line 66 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQbkOo-oisURHIGP0Y1:-KQbkOo-oisURHIGP0Y2:bk6xy2t) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/cluster/kube-util.sh#L66)):*
> ```Shell
> 	# re-sourced so the change propogates to dependent computed values
> 	# (eg: MASTER_SG_NAME, NODE_SG_NAME, etc)
> 	source "${KUBE_ROOT}/cluster/aws/util.sh"
> ```

This hack is fine as long as it is contained in the federation part of the code. Let's make sure we don't leak this to non-federation code.

---

*[federation/cluster/common.sh, line 260 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQblE1e2yBrXy6B-dC-:-KQblE1e2yBrXy6B-dC0:b-q9lu27) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/federation/cluster/common.sh#L260)):*
> ```Shell
> 
>   if [[ "${IS_DNS_NAME:-}" == "true" ]];then
>       primary_cn="${MASTER_NAME}"
> ```

Why not "${1}" here? The goal is to make Certificate's CN as unique as possible. Using a constant, i.e. the master name defeats that. Let's not do that. 

---

*[federation/cluster/common.sh, line 261 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQblikUO6_wAvfJDcsO:-KQblikVxu-pe81j2G16:b-nul9rh) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/federation/cluster/common.sh#L261)):*
> ```Shell
>   if [[ "${IS_DNS_NAME:-}" == "true" ]];then
>       primary_cn="${MASTER_NAME}"
>       sans="DNS:${1}"
> ```

You still need master name as SANS (or probably not, both here and in the next case).

---

*[federation/manifests/federation-apiserver-deployment.yaml, line 26 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQbqz8ydpI6pwPuTMaO:-KQbqz8ydpI6pwPuTMaP:bstw3j9) ([raw file](https://github.com/kubernetes/kubernetes/blob/07556d5e1ae90a141a09c9958dfb875de4298143/federation/manifests/federation-apiserver-deployment.yaml#L26)):*
> ```YAML
>         - --service-cluster-ip-range={{.FEDERATION_SERVICE_CIDR}}
>         - --secure-port=443
>         - --advertise-address={{.FEDERATION_API_HOST}}
> ```

Can we set this to empty string in the cases where `FEDERATION_API_HOST` is not an IP? We could probably have another variable called `FEDERATION_API_HOST_IP` that could be set or unset depending on the type of the `FEDERATION_API_HOST` value.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-:-KQbrQynj5hJtEJ6Jkjy:b-30zn32)*
<!-- Sent from Reviewable.io -->

quinton-hoole commented:
@colhom The Federated Ingress failure is happening on GCE too, so I think it's a bug (possibly in the test).  The others look like DNS issues. 

I'd be happy to get this script reviewed and merged, even if some of the tests are still failing.  I think it will be easier to debug once Jenkins is running the tests continuously.

ixdy commented:
cc @david-mcmahon re: `build/release.sh` changes (I hope we can kill that off soon)

colhom commented:
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[build/release.sh, line 34 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQbjPYTGCBxRnRkWsHU:-KQc7D3XfZ75qUFfyW3t:b-3ksfi4) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/build/release.sh#L34)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> I am not entirely sure about this change, particularly about putting a large portion of this script behind an environment variable. What stops us from having a separate script that just has
> 
> ```
> kube::release::package_tarballs
> kube::release::package_hyperkube
> ```
> ?
> 
> @david-mcmahon @thockin for their opinions.
</details>
I agree we should probably just have a separate script that does this shortcut. If anyone doesn't object, I'll just yank that commit out and throw in a `build/package-artifacts.sh` which does , or something to that effect?

---

*[cluster/kube-util.sh, line 66 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQbkOo-oisURHIGP0Y1:-KQc7RWDboy5CzrzR9Sq:bwyl1mk) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/cluster/kube-util.sh#L66)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> This hack is fine as long as it is contained in the federation part of the code. Let's make sure we don't leak this to non-federation code.
</details>
The `set-federation-zone-vars` that we see here is only in the codepath when `FEDERATION=true`. If this is ever not true, we definitely need to fix it.

---

*[federation/cluster/common.sh, line 260 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQblE1e2yBrXy6B-dC-:-KQc7enr2pssD2z4xDgr:b-2wqaot) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/federation/cluster/common.sh#L260)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> Why not "${1}" here? The goal is to make Certificate's CN as unique as possible. Using a constant, i.e. the master name defeats that. Let's not do that. 
</details>
Agree entirely, though unfortunately Common Names are restricted to being 64 bytes long. In the case of AWS load balancers, `${1}` is a very, very long DNS hostname > 64 characters (rather than an IPv4 address in the GCE case) , so it can't be the common name. I agree we're abusing the spec, but the alternative is adding another layer of DNS indirection to this setup. 

---

*[federation/cluster/common.sh, line 261 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQblikUO6_wAvfJDcsO:-KQc8Lz3poFzR5-hez7o:b-m9z732) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/federation/cluster/common.sh#L261)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> You still need master name as SANS (or probably not, both here and in the next case).
</details>
In the `${1}` is a DNS name case, `${MASTER_NAME}` is the CN- so it doesn't need to be in the sans.

---

*[federation/manifests/federation-apiserver-deployment.yaml, line 26 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQbqz8ydpI6pwPuTMaO:-KQc8Vjzf_r8IX4ruoRt:b-z769mm) ([raw file](https://github.com/kubernetes/kubernetes/blob/07556d5e1ae90a141a09c9958dfb875de4298143/federation/manifests/federation-apiserver-deployment.yaml#L26)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> Can we set this to empty string in the cases where `FEDERATION_API_HOST` is not an IP? We could probably have another variable called `FEDERATION_API_HOST_IP` that could be set or unset depending on the type of the `FEDERATION_API_HOST` value.
</details>
Given that the API server is serving behind a load balancer which may or may not have an IP permanently associated with it, `--advertise-address` defaulting to the bind address is what we want in any case (leave it blank). What we should really do if we want to be correct is set the `--bind-address` to the podIP via the downward API, and then `advertise-address` will be correct as well.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908)*
<!-- Sent from Reviewable.io -->

colhom commented:
@madhusudancs you have the logs for two identical runs in your email.

madhusudancs commented:
Context for folks (other than Colin and me) following this PR. @colhom and I had a discussion about this over email and Colin is going to fix the route53/IAM issue due to which the service tests are failing. We will get this PR merged after that.

madhusudancs commented:
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[build/release.sh, line 34 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQbjPYTGCBxRnRkWsHU:-KQd9t5HbGLVeKSYWgQh:b-pag5jf) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/build/release.sh#L34)):*
<details><summary><i>Previously, colhom wrote…</i></summary>
> I agree we should probably just have a separate script that does this shortcut. If anyone doesn't object, I'll just yank that commit out and throw in a `build/package-artifacts.sh` which does , or something to that effect?
</details>
In any case, please separate that out into a new PR (and revert these changes here) so that this PR can get reviewed and merged independently.

---

*[cluster/kube-util.sh, line 66 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQbkOo-oisURHIGP0Y1:-KQdAGVp4yZnCNT4UAfO:b8dlr65) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/cluster/kube-util.sh#L66)):*
<details><summary><i>Previously, colhom wrote…</i></summary>
> The `set-federation-zone-vars` that we see here is only in the codepath when `FEDERATION=true`. If this is ever not true, we definitely need to fix it.
</details>
Sure. It was just a caution or more of a self-note.

---

*[federation/cluster/common.sh, line 261 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQblikUO6_wAvfJDcsO:-KQdAmHx4ZVSrz3cVbNE:b-mq3pfs) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/federation/cluster/common.sh#L261)):*
<details><summary><i>Previously, colhom wrote…</i></summary>
> In the `${1}` is a DNS name case, `${MASTER_NAME}` is the CN- so it doesn't need to be in the sans.
</details>
We can take the first 54 characters of the DNS name (64 - 10 for the Unix date)?

---

*[federation/manifests/federation-apiserver-deployment.yaml, line 26 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQbqz8ydpI6pwPuTMaO:-KQdBidSUaIn5_UqY27k:bvaq98i) ([raw file](https://github.com/kubernetes/kubernetes/blob/07556d5e1ae90a141a09c9958dfb875de4298143/federation/manifests/federation-apiserver-deployment.yaml#L26)):*
<details><summary><i>Previously, colhom wrote…</i></summary>
> Given that the API server is serving behind a load balancer which may or may not have an IP permanently associated with it, `--advertise-address` defaulting to the bind address is what we want in any case (leave it blank). What we should really do if we want to be correct is set the `--bind-address` to the podIP via the downward API, and then `advertise-address` will be correct as well.
</details>
@nikhiljindal can add more context here. But in any case, I talked to him today and he said that this IP is also used in the documentation, so it definitely can't be pod IP. But I agree that `--bind-address` should be the Pod IP.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[federation/cluster/common.sh, line 260 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQblE1e2yBrXy6B-dC-:-KQdDVaDynBONZ-v8KVw:b-mq3pfs) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/federation/cluster/common.sh#L260)):*
<details><summary><i>Previously, colhom wrote…</i></summary>
> Agree entirely, though unfortunately Common Names are restricted to being 64 bytes long. In the case of AWS load balancers, `${1}` is a very, very long DNS hostname > 64 characters (rather than an IPv4 address in the GCE case) , so it can't be the common name. I agree we're abusing the spec, but the alternative is adding another layer of DNS indirection to this setup. 
</details>
We can take the first 54 characters of the DNS name (64 - 10 for the Unix date)?

---

*[federation/cluster/common.sh, line 261 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQblikUO6_wAvfJDcsO:-KQdCl1jeC_vuJw1txKb:bkx29ap) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/federation/cluster/common.sh#L261)):*
> We can take the first 54 characters of the DNS name (64 - 10 for the Unix date)?

This comment wasn't supposed to be here. It was for the previous one. My fault, I scrolled too fast or something.

Anyway.

> In the ${1} is a DNS name case, ${MASTER_NAME} is the CN- so it doesn't need to be in the sans.

`MASTER_NAME=federation-apiserver` can't be the CA's CN from my comments above. You need something more unique than that. If you use the first 54 chars of an AWS LB address for CA's CN then `MASTER_NAME=federation-apiserver` can be used as the server's CN. Then I agree that it is not necessary to have `$MASTER_NAME` in `$SANS`. That also means, it is not necessary to have `$MASTER_NAME` in SANS below. It's pretty strange we had that already.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908)*
<!-- Sent from Reviewable.io -->

colhom commented:
```sh
Summarizing 4 Failures:

[Fail] [k8s.io] Federation events [Feature:Federation] [BeforeEach] Event objects should be created and deleted successfully
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:209

[Fail] [k8s.io] [Feature:Federation] Federated Services DNS non-local federated service [It] should be able to discover a non-local federated service
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:368

[Fail] [k8s.io] Federated ingresses [Feature:Federation] Federated Ingresses [It] should create and update matching ingresses in underlying clusters
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federated-ingress.go:289

[Fail] [k8s.io] [Feature:Federation] Federated Services DNS [It] should be able to discover a federated service
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:368

Ran 16 of 383 Specs in 1543.772 seconds
FAIL! -- 12 Passed | 4 Failed | 1 Pending | 366 Skipped --- FAIL: TestE2E (1543.79s)
FAIL
```

The Route53 permission problem has been fixed, one less failure.

colhom commented:
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[cluster/kube-util.sh, line 66 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQbkOo-oisURHIGP0Y1:-KQdUHPzxR1R0uva4MRo:b-du5089) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/cluster/kube-util.sh#L66)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> Sure. It was just a caution or more of a self-note.
</details>
This is how you enforce invariants in bash. Leave yourself and your colleagues notes. TODO: rewrite all of this in a real language.

---

*[federation/cluster/common.sh, line 261 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQblikUO6_wAvfJDcsO:-KQdSgm9WlPR-2qdujFX:bvb0y4g) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/federation/cluster/common.sh#L261)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> > We can take the first 54 characters of the DNS name (64 - 10 for the Unix date)?
> 
> This comment wasn't supposed to be here. It was for the previous one. My fault, I scrolled too fast or something.
> 
> Anyway.
> 
> > In the ${1} is a DNS name case, ${MASTER_NAME} is the CN- so it doesn't need to be in the sans.
> 
> `MASTER_NAME=federation-apiserver` can't be the CA's CN from my comments above. You need something more unique than that. If you use the first 54 chars of an AWS LB address for CA's CN then `MASTER_NAME=federation-apiserver` can be used as the server's CN. Then I agree that it is not necessary to have `$MASTER_NAME` in `$SANS`. That also means, it is not necessary to have `$MASTER_NAME` in SANS below. It's pretty strange we had that already.
> 
</details>
Taking any sub-set of the elb name is potential for collisions, even with the timestamp (clock skew/errors on ec2 instances can be brutal). We should just hash the entire ELB name if we're going this route. The fact that you can't generate the full DNS name from the Common Name shouldn't be problematic.

---

*[federation/manifests/federation-apiserver-deployment.yaml, line 26 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQbqz8ydpI6pwPuTMaO:-KQdRbTIzXoHmsRHZk0a:b-l78j34) ([raw file](https://github.com/kubernetes/kubernetes/blob/07556d5e1ae90a141a09c9958dfb875de4298143/federation/manifests/federation-apiserver-deployment.yaml#L26)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> @nikhiljindal can add more context here. But in any case, I talked to him today and he said that this IP is also used in the documentation, so it definitely can't be pod IP. But I agree that `--bind-address` should be the Pod IP.
</details>
The fact that the API server unilaterally expects to be able to advertise a static IP address to __API clients__ is inherently broken. What about API servers that live behind ELBs, which are only transiently associated with IP addresses? 

Precluding any immediate changes to the generic API server package, the only IP addresses that even remotely make sense in that case for `--advertise-ip`:

* private IP of ec2 instance hosting API server (pod)
* podIP (if it's a k8s pod)

I say we make `--advertise` and `--bind` both `podIP` and file an issue about this.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
@colhom `Ingress` one is a known failure. It is failing on GCE as well. `Event` one looks like a flake. It wasn't there in your previous run. So I guess it has passed before so we can merge this PR without blocking on that and debug afterwards. The two `Service` failures are likely due to the unregistered domain name problem we have discussed in the email. If that's the problem we don't have to block on that as well.

---

Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[cluster/kube-util.sh, line 66 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQbkOo-oisURHIGP0Y1:-KQdsXMVqIomEzXM0dLm:b-en5opf) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/cluster/kube-util.sh#L66)):*
>  TODO: rewrite all of this in a real language.

I don't know if you have already noticed, but this is already happening - PR #30615, PR #31856 and others.

---

*[federation/cluster/common.sh, line 261 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQblikUO6_wAvfJDcsO:-KQdtGcoecMBqbFV3eeu:bo2g7nt) ([raw file](https://github.com/kubernetes/kubernetes/blob/846fdf457025b36530d08ffd8045e0aa113df1b5/federation/cluster/common.sh#L261)):*
> Taking any sub-set of the elb name is potential for collisions, even with the timestamp (clock skew/errors on ec2 instances can be brutal). We should just hash the entire ELB name if we're going this route.

Hash makes sense, let's do that. 

>  The fact that you can't generate the full DNS name from the Common Name shouldn't be problematic.

That's fine.

---

*[federation/manifests/federation-apiserver-deployment.yaml, line 26 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQbqz8ydpI6pwPuTMaO:-KQdu3Att-kPXjjPgT7-:b2gf25w) ([raw file](https://github.com/kubernetes/kubernetes/blob/07556d5e1ae90a141a09c9958dfb875de4298143/federation/manifests/federation-apiserver-deployment.yaml#L26)):*
> The fact that the API server unilaterally expects to be able to advertise a static IP address to API clients is inherently broken. What about API servers that live behind ELBs, which are only transiently associated with IP addresses?

Agreed. @nikhiljindal and I briefly discussed this today. It should probably be a string instead of an IP.

> Precluding any immediate changes to the generic API server package, the only IP addresses that even remotely make sense in that case for --advertise-ip:
> podIP (if it's a k8s pod)

I don't think so. If we are displaying `--advertise-address` in any user visible way, it cannot be a pod address. It is not just about pod address not being stable, it is also not routable outside the cluster.

The thing that makes sense is to set it when we have a static address and leave it out when we don't. And then we can nag @nikhiljindal to fix the flag ;-)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-:-KQdpbapRx6pFQXJlm0p:brpej7k)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions.

---

*[cluster/aws/util.sh, line 724 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQdx2-ITFMV_2y-C0qP:-KQdx2-Kz7mwba-88j31:bpzlwms) ([raw file](https://github.com/kubernetes/kubernetes/blob/07556d5e1ae90a141a09c9958dfb875de4298143/cluster/aws/util.sh#L724)):*
> ```Shell
> # Creates the IAM roles (if they do not already exist)
> function ensure-iam-profiles {
>   aws iam get-instance-profile --instance-profile-name ${IAM_PROFILE_MASTER} || {
> ```

I am not an expert in this part of the code. 

@justinsb could you please help review the changes to this file?

---

*[cluster/aws/util.sh, line 803 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQdxcGGcdrisiZqjioo:-KQdxcGHl8qA68AmInfC:b2ccwj1) ([raw file](https://github.com/kubernetes/kubernetes/blob/614e01f09bc9ebbb38863aa559a9aa05d66a8516/cluster/aws/util.sh#L803)):*
> ```Shell
> function delete-iam-profiles {
>   for iam_profile_name in ${IAM_PROFILE_MASTER} ${IAM_PROFILE_NODE};do
>       echo "Removing role from instance profile: ${iam_profile_name}"
> ```

Please fix the indentation. It looks like rest of the file uses a 2-space indentation.

---

*[cluster/aws/util.sh, line 824 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-KQdxeaX9WONyQF3kXyV:-KQdxeaX9WONyQF3kXyW:bbu5wl8) ([raw file](https://github.com/kubernetes/kubernetes/blob/614e01f09bc9ebbb38863aa559a9aa05d66a8516/cluster/aws/util.sh#L824)):*
> ```Shell
>     local -r errMsg=$($@ 2>&1 > $LOG)
>     if [[ "$errMsg" == "" ]];then
> 	return
> ```

Looks like a tab here.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908)*
<!-- Sent from Reviewable.io -->

colhom commented:
@madhusudancs are you shepherd on this? I'll wait till @justinsb reviews the IAM stuff, fixup and then I'll remove the `build/release.sh` hackery commit and squash up for merge?

madhusudancs commented:
@colhom yes, I will shepherd this. I have also pinged @justinsb. Sorry about the delay.

---

Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31908#-:-KR5JbmTnTJJ202e0-Rt:b-95zwpb)*
<!-- Sent from Reviewable.io -->

justinsb commented on cluster/aws/util.sh:
```diff
@@ -411,9 +411,9 @@ function create-security-group {
-	  add-tag $sgid KubernetesCluster ${CLUSTER_ID}
+          echo "Creating security group ${name}."
+          sgid=$($AWS_CMD create-security-group --group-name "${name}" --description "${description}" --vpc-id "${VPC_ID}" --query GroupId)
+          add-tag $sgid KubernetesCluster ${CLUSTER_ID}
```
Is this just a spacing change?  Maybe remove from the PR, or match the 2-space approach we seem to use in this file?

colhom replied:
This is normalizing some previously-merged lines from tab to 2-space (My fault from previous PRs). I ran an emacs macro over the relevant files, so they should all be uniform after this PR.

justinsb commented on cluster/aws/util.sh:
```diff
@@ -721,14 +723,11 @@ function create-iam-profile {
 
 # Creates the IAM roles (if they do not already exist)
 function ensure-iam-profiles {
-  aws iam get-instance-profile --instance-profile-name ${IAM_PROFILE_MASTER} || {
```
What if the instance profile already exists?

colhom replied:
Before the cluster comes up, the new function `delete-iam-profiles` is run to ensure any pre-existing iam roles/profiles/relationships are cleaned up before this point.

madhusudancs replied:
Isn't that dangerous? I am not entirely sure if we want to delete it without any warning or an explicit confirmation from the user. Note that, kube-up/kube-down are not just used in tests. It is used in real production setups by users.

colhom replied:
@madhusudancs these IAM roles are created entirely by the `kube-up` script, so it's reasonable that they are destroyed by the script as well. This is the modus operandi for resource allocation- if we create it automatically, we clean it up automatically as well.

madhusudancs replied:
> This is the modus operandi for resource allocation- if we create it automatically, we clean it up automatically as well.

This isn't entirely true. We do take a more conservative approach at several places and that's the most reasonable thing to do. Examples being: Federated Services, Federated Replica Sets, Kubernetes Deployments and in various other places.

Moreover, what makes it reasonable to assume that these resources were created entirely by `kube-up` script? Do we add an annotation or something that we check against before deleting it? Is there a hash that's hard to reproduce?

colhom replied:
> examples being: Federated Services, Federated Replica Sets, Kubernetes Deployments and in various other places.

All these examples are within the e2e framework? This file deals solely with provisioning AWS resources, not Kubernetes api objects. If you look at any other type of AWS resource which is created in `kube-up.sh`, it's also destroyed in `kube-down.sh`.

If we don't do this, then any changes made to the IAM roles/instance profiles will not be propagate properly to e2e tests. The IAM roles/profiles will have to be managed manually.

Finally, why are we so concerned about deleted these IAM roles? They're created and used exclusively by the cluster (just like all the other AWS resources), so I don't know why we're so concerned about retaining them. These IAM objects exist simply to allow the ec2 instances to access the AWS API- can you state explicitly the rationale behind not deleting them when the cluster is torn down?

madhusudancs replied:
> All these examples are within the e2e framework? 

No and so does this file. This file isn't used only in our test framework. There are real users, using `kube-up` in production today. And they could be using it for multiple clusters.

> If you look at any other type of AWS resource which is created in kube-up.sh, it's also destroyed in kube-down.sh.

Yeah, but this isn't deleting in `kube-down.sh`. So it is not exactly deleting what it created. It is reversing the model. It might accidentally delete what it did not create when it comes up the first time. As far as I see, it is deleting in `kube-up.sh`. Am I wrong/missing something here?

> If we don't do this, then any changes made to the IAM roles/instance profiles will not be propagate properly to e2e tests.

This smells like something that must be done in an e2e specific environment, not here, given that users are going to use in their production environments.

> Finally, why are we so concerned about deleted these IAM roles? They're created and used exclusively by the cluster (just like all the other AWS resources),

Yeah, but that assumes that there is only one cluster. There could another cluster with the same name. We can't simply go ahead and delete its IAM roles without any warning. 

> can you state explicitly the rationale behind not deleting them when the cluster is torn down?

The rationale is above. But stating it again. We can't delete a resource, as part of the bootstrap, unless we are 100% sure we created them. There is nothing here to guarantee that. It is not even close. We are deleting without any checks, without any warnings in place.

colhom replied:
Ok, I understand your concern, thanks for explaining.

https://github.com/kubernetes/kubernetes/blob/master/cluster/aws/util.sh#L1319-L1339

You'll see that currently the "checks" applied to resource cleanup consist of two unique values:
*  `CLUSTER_ID`, which is traditionally defined to be `INSTANCE_PREFIX` in all the configs I've seen
* `VPC_ID`, which obtained by looking up the ID for `VPC_NAME` (provided in config)

If I applied the same tagging/ deletion check logic to the IAM roles/profiles, that would be consistent with how everything else works.

Sound like a plan?

madhusudancs replied:
Sounds good.

And these things should be configured to have test specific values (strings) in test configs.

justinsb commented on cluster/aws/util.sh:
```diff
@@ -815,12 +854,12 @@ function vpc-setup {
+          $AWS_CMD modify-vpc-attribute --vpc-id $VPC_ID --enable-dns-support '{"Value": true}' > $LOG
+          $AWS_CMD modify-vpc-attribute --vpc-id $VPC_ID --enable-dns-hostnames '{"Value": true}' > $LOG
+          add-tag $VPC_ID Name ${VPC_NAME}
+          add-tag $VPC_ID KubernetesCluster ${CLUSTER_ID}
```
Just spacing changes?

colhom replied:
Yep.

justinsb commented on cluster/aws/util.sh:
```diff
@@ -880,9 +919,9 @@ function kube-up {
-	  $AWS_CMD attach-internet-gateway --internet-gateway-id $IGW_ID --vpc-id $VPC_ID > $LOG
+          echo "Creating Internet Gateway."
+          IGW_ID=$($AWS_CMD create-internet-gateway --query InternetGateway.InternetGatewayId)
+          $AWS_CMD attach-internet-gateway --internet-gateway-id $IGW_ID --vpc-id $VPC_ID > $LOG
```
Just spacing?

colhom replied:
Yep.

justinsb commented on cluster/aws/config-test.sh:
```diff
@@ -56,8 +56,8 @@ CONFIG_CONTEXT="${KUBE_CONFIG_CONTEXT:-aws_${INSTANCE_PREFIX}}"
 AWS_SSH_KEY=${AWS_SSH_KEY:-$HOME/.ssh/kube_aws_rsa}
-IAM_PROFILE_MASTER="kubernetes-master"
-IAM_PROFILE_NODE="kubernetes-minion"
+IAM_PROFILE_MASTER="kubernetes-master-${INSTANCE_PREFIX}"
```
Have we previously changed config-default?

justinsb commented on cluster/kube-util.sh:
```diff
@@ -43,22 +43,29 @@ function set-federation-zone-vars {
+        # gcloud has a 61 character limit, and for firewall rules this
+        # prefix gets appended to itself, with some extra information
+        # need tot keep it short
+        export KUBE_GCE_INSTANCE_PREFIX="${USER}-${zone}"
```
Just spacing?

colhom replied:
Yep.

justinsb commented on cluster/kube-util.sh:
```diff
@@ -43,22 +43,29 @@ function set-federation-zone-vars {
+        # we need to make sure the config-xxx.sh file is
+        # re-sourced so the change propogates to dependent computed values
+        # (eg: MASTER_SG_NAME, NODE_SG_NAME, etc)
+        source "${KUBE_ROOT}/cluster/aws/util.sh"
```
Hmmm... why does GCE not have the same problem?

colhom replied:
I honestly can't say for sure, I'd have to re-read that util file. As happened here, it's stil possible we're stomping some resource definition by launching multiple clusters and just not noticing :-1:  

IIRC, GCE doesn't do cross-zone namespacing with the components we use, so we get away with conflicting names across our cluster set. We require 1 or 0 clusters per zone (for all providers), so it works out in the GCE case.

colhom commented:
Side note- spacing is still messed up. I'm going to respond to these review items, fix the spacing, drop the `build/release/sh commit and then rebase/squash on master.

colhom commented on cluster/kube-util.sh:
```diff
@@ -38,27 +38,34 @@ fi
 # Should NOT be called within the global scope, unless setting the desired global zone vars
 # This function is currently NOT USED in the global scope
 function set-federation-zone-vars {
```
I wrote most of this file, so I felt compelled to fix the indent problem on the entire function.

madhusudancs replied:
Thanks!

colhom commented:
@madhusudancs PTAL

colhom commented:
@madhusudancs @quinton-hoole I have a weird e2e failure that popped up on a run today.

```sh
reflector.go:214] k8s.io/kubernetes/federation/pkg/federation-controller/util/federated_informer.go:275: Failed to list *v1beta1.Cluster: the server could not find the requested resource
```
This line was repeated thousands of times- I feel like this is related to cluster graduating to the beta api?

madhusudancs commented:
@colhom Have you rebased to master recently? What are the test failure messages you are seeing?

Our GCE tests have started failing too while we weren't looking at them for a few days when they were out of quota - https://k8s-testgrid.appspot.com/google-gce#gce-federation. @nikhiljindal is looking into it. Do AWS errors look like this - https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubernetes-e2e-gce-federation/4026?

colhom commented:
@madhusudancs 
```sh
Summarizing 10 Failures:

[Fail] [k8s.io] [Feature:Federation] Federated Services [BeforeEach] DNS non-local federated service should be able to discover a non-local federated service 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:83

[Fail] [k8s.io] Federation secrets [Feature:Federation] Secret objects [BeforeEach] should be created and deleted successfully 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:94

[Fail] [k8s.io] [Feature:Federation] Federated Services [BeforeEach] DNS should be able to discover a federated service 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:94

[Fail] [k8s.io] [Feature:Federation] Federated Services [BeforeEach] DNS non-local federated service [Slow] missing local service should never find DNS entries for a missing local service 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:94

[Fail] [k8s.io] Federation replicasets [Feature:Federation] Federated ReplicaSet [BeforeEach] should create and update matching replicasets in underling clusters 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:94

[Fail] [k8s.io] Federation apiserver [Feature:Federation] Cluster objects [It] should be created and deleted successfully 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:94

[Fail] [k8s.io] Federation namespace [Feature:Federation] Namespace objects [BeforeEach] should be created and deleted successfully 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util-14.go:88

[Fail] [k8s.io] [Feature:Federation] Federated Services [BeforeEach] Service creation should create matching services in underlying clusters 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:94

[Fail] [k8s.io] [Feature:Federation] Federated Services [BeforeEach] Service creation should succeed 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:94

[Fail] [k8s.io] Federated ingresses [Feature:Federation] Federated Ingresses [BeforeEach] should create and update matching ingresses in underlying clusters 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:94

Ran 17 of 389 Specs in 2865.780 seconds
FAIL! -- 7 Passed | 10 Failed | 1 Pending | 371 Skipped --- FAIL: TestE2E (2865.80s)
FAIL
```

I had rebased yesterday morning, I'll do so again now and see what happens.

quinton-hoole commented:
@colhom Are you blocked on anything here?  @madhusudancs and I are pretty snowed under with getting v1.4 out the door this week, but can help you debug as soon as that's done.  I don't consider this a v1.4 blocker.

colhom commented:
@quinton-hoole as of the end of last week, we're still experiencing e2e failures- I believe the root cause is spelled out in the log line here.:

```sh
reflector.go:214] k8s.io/kubernetes/federation/pkg/federation-controller/util/federated_informer.go:275: Failed to list *v1beta1.Cluster: the server could not find the requested resource
```

Seems that there is a mismatch between the controller's and apiserver's supported apis. In the mean time, I'll double-check the versions of all the federation control plane pods and make sure that this isn't some esoteric artifact mismatch symptom (api server/controller version mismatch or something screwy like that)

madhusudancs commented:
@colhom sorry, I did not reply to you earlier. A bug in `clientcmd` code is causing all federation tests to fail. You are seeing the same errors we are seeing on GCE. There is a PR out there to fix it - https://github.com/kubernetes/kubernetes/pull/32438#issuecomment-246170244. So hopefully we can make progress here after that fix is in.

Also, in the meantime I will summon @justinsb for help with one more round of reviews :)

colhom commented:
@madhusudancs I've rebased on master after #32613 merged- seems to have rectified the specific api version issue but there are still failures (check your email).

I have kubernetes/test-infra#577 up to add this functionality to the official jenkins config, and you'll be able to debug further.

madhusudancs commented on federation/manifests/federation-apiserver-deployment.yaml:
```diff
@@ -23,7 +23,6 @@ spec:
         - --etcd-servers=http://localhost:2379
         - --service-cluster-ip-range={{.FEDERATION_SERVICE_CIDR}}
         - --secure-port=443
-        - --advertise-address={{.FEDERATION_API_HOST}}
```
Following up from the previous comments. Can we please bring this back? At least on the providers where we do have an IP address.

colhom replied:
The flag seems inherently misguided, in that it expects that the advertise address for the federation apiserver is an IP. I'd say the external address for the federation API server being a static IP address is a special case (local dev and GCE), rather than the norm.

madhusudancs replied:
>  I'd say the external address for the federation API server being a static IP address is a special case (local dev and GCE)

Yes. But in that special case it is an IP and it can be advertised that way.

Also, why unnecessarily break the existing behavior?

colhom replied:
What existing behavior are we breaking? As far as I know, there is no functionality in the federation control plane which is affected by the `--advertise-address` simply defaulting to the bind address. If there were, we should fix it so it will work on all platforms, rather than just GCE/local dev (rather than make this change here).

madhusudancs replied:
> What existing behavior are we breaking? As far as I know, there is no functionality in the federation control plane which is affected by the --advertise-address 

There is, apparently, and I mentioned it in this code review earlier. We do use this field for doc generation.

> simply defaulting to the bind address. 

And defaulting to bind address is not even correct. If you have 2 pods for a federation API server, what's the right address to advertise?

> If there were, we should fix it so it will work on all platforms, rather than just GCE/local dev (rather than make this change here).

There is no second question about that. But that doesn't mean we need to unnecessarily change this.

colhom replied:
Will do. I've filed a bug in the meantime #34043

madhusudancs replied:
Cool!

madhusudancs reviewed (changes requested):
@colhom just two comments and then we are good to go. Sorry about the delay :(

colhom reviewed (commented):
@madhusudancs thanks for dealing with my stubbornness, will fix.

madhusudancs reviewed (changes requested):
@colhom Let me know when you update this PR with what we have discussed. Let's get this in soon!

Commit:
fixup: iam roles are now uniquely identified by CLUSTER_ID + VPC_ID

this is the same criteria that is currently applied for ascertaining ownership other resources ( eg: security groups, which are also deleted in a similar fashion)

We are now respecting existing cluster namespace isolation rules.

Commit:
fixup! set advertise address if not ip address

colhom commented:
@madhusudancs sorry for the delay. The review items have been hit and the commits rebased on master. I'm running this through a final e2e run over on my side to verify the changes are sane.

colhom commented:
```sh
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS

Summarizing 5 Failures:

[Fail] [k8s.io] [Feature:Federation] [BeforeEach] Federated Services DNS should be able to discover a federated service
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:221

[Fail] [k8s.io] [Feature:Federation] Federation API server authentication [It] should not accept cluster resources when the client has invalid authentication credentials 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-authn.go:63

[Fail] [k8s.io] [Feature:Federation] Federation API server authentication [It] should not accept cluster resources when the client has no authentication credentials 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-authn.go:77

[Fail] [k8s.io] Federated ingresses [Feature:Federation] Federated Ingresses Ingress connectivity and DNS [It] should be able to connect to a federated ingress via its load balancer 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federated-ingress.go:334

[Fail] [k8s.io] [Feature:Federation] Federated Services DNS non-local federated service [It] should be able to discover a non-local federated service 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-util.go:447

Ran 17 of 409 Specs in 1979.515 seconds
FAIL! -- 12 Passed | 5 Failed | 1 Pending | 391 Skipped --- FAIL: TestE2E (1979.53s)
FAIL
```

^^ latest e2e run. I think we should merge and iterate on these failures now.

quinton-hoole commented:
@colhom Awesome.  At least a few of those are probably expected to fail at this point:

1. "Federated Services DNS non-local federated service  ..."  is probably failing due to credentials for AWS Route53 access not being set up correctly in your test.
2. "Federated Ingresses - Ingress connectivity..." is failing because Federated Ingress is not currently supported on AWS (or anywhere other than GCE and GKE).  So we need to add an appropriate Skip() clause to that test (that's an easy one-liner).
3. "Federated Services DNS should be able to discover a federated service" is probably failing for the same reason, and should also be skipped.

Merge it :-)
Q

madhusudancs commented:
@colhom cool! LGTM! Thanks for sticking with this and with the endless delays!

All the test failures are either known (i.e. happening on GCE too) or expected (Services due to real domain name). Let's get this in!

quinton-hoole commented:
Actually my point 3 is incorrect. We'll have to dig deeper into that one.

madhusudancs commented:
@k8s-bot ok to test

madhusudancs commented:
@colhom I think you need to sign the CNCF CLA to get this PR in.
